controller/deployer/strategies/main.go

```diff
@@ -7,22 +7,124 @@ import (
+type UnknownStrategyError struct {
+	Strategy string
+}
+
+func (e UnknownStrategyError) Error() string {
+	return fmt.Sprintf(`deployer: unknown strategy "%s"`, e.Strategy)
+}
+
+type Deploy struct {
+	*ct.Deployment
+	client        *controller.Client
+	deployEvents  chan<- ct.DeploymentEvent
+	jobEvents     chan *ct.JobEvent
+	serviceEvents chan *discoverd.Event
+	useJobEvents  map[string]struct{}
+	logger        log15.Logger
+}
+
+type PerformFunc func(d *Deploy) error
 
 var performFuncs = map[string]PerformFunc{
 	"all-at-once": allAtOnce,
 	"one-by-one":  oneByOne,
 }
 
-func Get(strategy string) (PerformFunc, error) {
-	if f, ok := performFuncs[strategy]; ok {
-		return f, nil
+func Perform(d *ct.Deployment, client *controller.Client, deployEvents chan<- ct.DeploymentEvent, logger log15.Logger) error {
+	log := logger.New("fn", "Perform", "deployment_id", d.ID, "app_id", d.AppID)
+
+	log.Info("validating deployment strategy")
+	performFunc, ok := performFuncs[d.Strategy]
+	if !ok {
+		err := UnknownStrategyError{d.Strategy}
+		log.Error("error validating deployment strategy", "err", err)
+		return err
 	}
-	return nil, fmt.Errorf("Unknown strategy '%s'!", strategy)
+
+	deploy := &Deploy{
+		Deployment:    d,
+		client:        client,
+		deployEvents:  deployEvents,
+		serviceEvents: make(chan *discoverd.Event),
+		useJobEvents:  make(map[string]struct{}),
+		logger:        logger.New("deployment_id", d.ID, "app_id", d.AppID),
+	}
+
+	log.Info("determining release services")
+	release, err := client.GetRelease(d.NewReleaseID)
+	if err != nil {
+		log.Error("error getting new release", "release_id", d.NewReleaseID, "err", err)
+		return err
+	}
+	for typ, proc := range release.Processes {
+		if proc.Service == "" {
+			log.Info(fmt.Sprintf("using job events for %s process type, no service defined", typ))
+			deploy.useJobEvents[typ] = struct{}{}
+			continue
+		}
+
+		log.Info(fmt.Sprintf("using service discovery for %s process type", typ), "service", proc.Service)
+		events := make(chan *discoverd.Event)
+		stream, err := discoverd.NewService(proc.Service).Watch(events)
+		if err != nil {
+			log.Error("error creating service discovery watcher", "service", proc.Service, "err", err)
+			return err
+		}
+		defer stream.Close()
+
+	outer:
+		for {
+			select {
+			case event, ok := <-events:
+				if !ok {
+					log.Error("error creating service discovery watcher, channel closed", "service", proc.Service)
+					return fmt.Errorf("deployer: could not create watcher for service: %s", proc.Service)
+				}
+				if event.Kind == discoverd.EventKindCurrent {
+					break outer
+				}
+			case <-time.After(5 * time.Second):
+				log.Error("error creating service discovery watcher, timeout reached", "service", proc.Service)
+				return fmt.Errorf("deployer: could not create watcher for service: %s", proc.Service)
+			}
+		}
+		go func() {
+			for {
+				event, ok := <-events
```

Comment:
Ok, so I am going to be changing this to something like:

```go
events := make(chan *discoverd.Event)
reconnect := func() {
	stream, err := discoverd.NewService(proc.Service).Watch(events)
}
go func() {
	for {
		event, ok := <-events
		if !ok {
			reconnect()
			continue
		}
	}
}()
```

I don't mind changing for now though.